tests/mammotion-cli.py

```diff
@@ -0,0 +1,1473 @@
+            
+            # Notification logic
+            current_activity_for_notification = summary_after_poll.get('activity')
+            current_progress_for_notification = summary_after_poll.get('work_progress_percent')
+
+            if self.last_activity_for_notification and current_activity_for_notification:
+                act_low = current_activity_for_notification.lower()
+                last_act_low = self.last_activity_for_notification.lower()
+                was_working = "working" in last_act_low or "mowing" in last_act_low
+                is_idle_or_completed = any(s in act_low for s in ["idle", "charging", "completed", "standby", "docked", "mode_ready"])
+
+                progress_val = None
+                if isinstance(self.last_progress_for_notification, (int, float)):
+                    progress_val = self.last_progress_for_notification
+                elif isinstance(self.last_progress_for_notification, str) and self.last_progress_for_notification.replace('%','').isdigit():
+                    try: progress_val = int(self.last_progress_for_notification.replace('%',''))
+                    except ValueError: pass
+                
+                if was_working and is_idle_or_completed and progress_val is not None and progress_val >= 95:
+                    logger.info(
+                        f"NOTIFICATION: Task likely completed! "
+                        f"Prev State: {self.last_activity_for_notification} ({self.last_progress_for_notification}%), "
+                        f"Current State: {current_activity_for_notification}"
+                    )
+            
+            self.last_activity_for_notification = current_activity_for_notification
+            self.last_progress_for_notification = current_progress_for_notification
+            
+            try:
+                await asyncio.wait_for(self.cli_exit_event.wait(), timeout=current_interval)
+            except asyncio.TimeoutError:
+                pass # Normal timeout, continue loop
+        logger.info("Periodic status monitor stopped.")
+
+    async def safe_command_execution(self, command_name: str, **kwargs):
+        """Execute command with proper error handling and logging, using exponential_backoff."""
+        if not self.active_mower_device:
+            raise MowerCommandError(f"Cannot execute '{command_name}': Mower not connected.")
+        
+        try:
+            # Use config for retries, or defaults from exponential_backoff if not set in config
+            max_retries = self.config.max_retries if hasattr(self.config, 'max_retries') else Constants.DEFAULT_MAX_RETRIES
+            initial_delay = self.config.initial_retry_delay if hasattr(self.config, 'initial_retry_delay') else Constants.DEFAULT_INITIAL_RETRY_DELAY
+            max_delay = self.config.max_retry_delay if hasattr(self.config, 'max_retry_delay') else Constants.DEFAULT_MAX_RETRY_DELAY
+
+            result = await exponential_backoff(
```

Comment:
commands to the gateway already use exponential backoff, so you're just duplicating here.